pkg/server/cloud/gce/instances.go

```diff
@@ -0,0 +1,445 @@
+	labels := c.getInstanceLabels(node.Name)
+	networkInterfaces := c.getInstanceNetworkSpec(node.Spec.Resources.PrivateIPOnly)
+	instanceType := c.getInstanceTypeURL(node.Spec.InstanceType)
+	spec := &compute.Instance{
+		Disks:             disks,
+		Labels:            labels,
+		MachineType:       instanceType,
+		Name:              name,
+		NetworkInterfaces: networkInterfaces,
+		Tags: &compute.Tags{
+			Items: c.bootSecurityGroupIDs,
+		},
+		Metadata: &compute.Metadata{
+			Items: []*compute.MetadataItems{
+				{
+					Key:   "user-data",
+					Value: &mds,
+				},
+			},
+		},
+	}
+	if node.Spec.Spot {
+		ar := false
+		spec.Scheduling = &compute.Scheduling{
+			AutomaticRestart:  &ar,
+			OnHostMaintenance: "TERMINATE",
+			Preemptible:       true,
+		}
+	}
+	return spec, nil
+}
+
+func (c *gceClient) StartNode(node *api.Node, metadata string) (*cloud.StartNodeResult, error) {
+	klog.V(2).Infof("Starting instance for node: %v", node)
+	spec, err := c.createInstanceSpec(node, metadata)
+	if err != nil {
+		return nil, err
+	}
+	klog.V(2).Infof("Starting node with security groups: %v subnet: '%s'",
+		c.bootSecurityGroupIDs, c.subnetName)
+	op, err := c.service.Instances.Insert(c.projectID, c.zone, spec).Do()
+	if err != nil {
+		return nil, util.WrapError(err, "startup error")
+	}
+	if err := c.waitOnOperation(op.Name, c.getZoneOperation); err != nil {
+		return nil, err
+	}
+	// Todo: catch and convert errors to notify us of
+	// out of capacity errors or invalid machine types
+	// see pkg/server/cloud/aws/instances.StartNode()
+	startResult := &cloud.StartNodeResult{
+		InstanceID:       spec.Name,
+		AvailabilityZone: c.zone,
+	}
+	return startResult, nil
+}
+
+func (c *gceClient) StartSpotNode(node *api.Node, metadata string) (*cloud.StartNodeResult, error) {
+	klog.V(2).Infof("Starting instance for node: %v", node)
+	spec, err := c.createInstanceSpec(node, metadata)
+	if err != nil {
+		return nil, err
+	}
+	klog.V(2).Infof("Starting node with security groups: %v subnet: '%s'",
+		c.bootSecurityGroupIDs, c.subnetName)
+	op, err := c.service.Instances.Insert(c.projectID, c.zone, spec).Do()
+	if err != nil {
+		return nil, util.WrapError(err, "startup error")
+	}
+	// Todo: catch and convert errors to notify us of
+	// out of capacity errors or invalid machine types
+	// see pkg/server/cloud/aws/instances.StartNode()
+	if err := c.waitOnOperation(op.Name, c.getZoneOperation); err != nil {
+		return nil, err
+	}
+	startResult := &cloud.StartNodeResult{
+		InstanceID:       spec.Name,
+		AvailabilityZone: c.zone,
+	}
+	return startResult, nil
+}
+
+func (c *gceClient) WaitForRunning(node *api.Node) ([]api.NetworkAddress, error) {
+	for {
+		status, err := c.getInstanceStatus(node.Status.InstanceID)
+		if err != nil {
+			klog.Errorf("Error waiting for instance to start: %v", err)
+			return nil, err
+		}
+
+		klog.V(5).Infof("status: %s", status)
+		if status == statusInstanceRunning {
+			break
+		}
+		time.Sleep(10 * time.Second)
```

Comment:
Two things:
- 10s is probably a bit long - can we reduce it to something like 5s? With 10s we might wait for an extra 10s if the instance comes up right after the last instance status query.
- Shouldn't we add a timeout? What happens when we start the instance, but it crashes/terminated outside of Kip, and never gets to running? Will we get stuck in this loop?

Reply:
1. I have decreased sleep time to 5 seconds in d7bc5a7
2. I believe @justnoise  is going to take a look at the timeout situation.

Reply:
Fixed this up in https://github.com/elotl/kip/commit/b4f0e5cd7bc7a7ccbb3ebc3ca0b53ecb22fd1e19